src/libraries/System.Private.CoreLib/src/System/Number.NumberToFloatingPointBits.cs

```diff
@@ -498,57 +1262,80 @@ private static uint NumberToSingleFloatingPointBits(ref NumberBuffer number, in
-                    result *= scale;
+                    // It is only valid to do this optimization for single-precision floating-point
+                    // values since we can lose some of the mantissa bits and would return the
+                    // wrong value when upcasting to double.
+
+                    float result = DigitsToUInt32(src, (int)(totalDigits));
+                    float scale = s_Pow10SingleTable[fastExponent];
+
+                    if (fractionalDigitsPresent != 0)
+                    {
+                        result /= scale;
+                    }
+                    else
+                    {
+                        result *= scale;
+                    }
+
+                    return BitConverter.SingleToUInt32Bits(result);
                 }
 
-                return BitConverter.SingleToUInt32Bits(result);
-            }
+                // When the number of significant digits is less than or equal to MaxMantissaFastPath and the
+                // scale is less than or equal to MaxExponentFastPath, we can take some shortcuts and just rely
+                // on floating-point arithmetic to compute the correct result. This is
+                // because each floating-point precision values allows us to exactly represent
+                // different whole integers and certain powers of 10, depending on the underlying
+                // formats exact range. Additionally, IEEE operations dictate that the result is
+                // computed to the infinitely precise result and then rounded, which means that
+                // we can rely on it to produce the correct result when both inputs are exact.
+                // This is known as Clinger's fast path
 
-            if ((totalDigits <= 15) && (fastExponent <= 22))
-            {
-                double result = DigitsToUInt64(src, (int)(totalDigits));
-                double scale = s_Pow10DoubleTable[fastExponent];
+                ulong mantissa = DigitsToUInt64(src, (int)(totalDigits));
 
-                if (fractionalDigitsPresent != 0)
+                if ((mantissa <= info.MaxMantissaFastPath) && (fastExponent <= info.MaxExponentFastPath))
                 {
-                    result /= scale;
+                    double mantissa_d = mantissa;
+                    double scale = s_Pow10DoubleTable[fastExponent];
+
+                    if (fractionalDigitsPresent != 0)
+                    {
+                        mantissa_d /= scale;
+                    }
+                    else
+                    {
+                        mantissa_d *= scale;
+                    }
+
+                    return BitConverter.SingleToUInt32Bits((float)(mantissa_d));
                 }
-                else
+
+                // Number Parsing at a Gigabyte per Second, Software: Practice and Experience 51(8), 2021
+                // https://arxiv.org/abs/2101.11408
+
+                var am = ComputeFloat(exponent, mantissa, info);
```

Comment:
Here in other places please replace `var` with explicit type, according to our [coding-style](https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md) we use it only when the type is explicitly named on the right-hand side

Reply:
ok, i'll make the change